obelisk/cpp/obelisk_cpp/include/obelisk_mujoco_sim_robot.h

```diff
@@ -164,6 +164,62 @@ namespace obelisk {
 
             data_ = mj_makeData(model_);
 
+            // Create the rendering thread
+            std::thread rendering_thread(std::bind(&ObeliskMujocoRobot::SimRender, this));
+
+            mujoco_setup_ = true;
+
+            while (!activation_complete_) {
+                std::this_thread::sleep_for(std::chrono::milliseconds(3));
+            }
+            RCLCPP_INFO_STREAM(this->get_logger(), "Starting Mujoco simulation loop.");
+
+            // Set the initial condition based on a keyframe
+            RCLCPP_INFO_STREAM(this->get_logger(), "Found " << this->model_->nkey << " Mujoco keyframes.");
+            for (int i = 0; i < this->model_->nkey; i++) {
+                std::string potential_keyframe(this->model_->names + this->model_->name_keyadr[i]);
+                if (potential_keyframe == this->get_parameter("ic_keyframe").as_string()) {
+                    RCLCPP_INFO_STREAM(this->get_logger(),
+                                       "Setting initial condition to keyframe: " << potential_keyframe);
+                    mju_copy(this->data_->qpos, &this->model_->key_qpos[i * this->model_->nq], this->model_->nq);
+                    mju_copy(this->data_->qvel, &this->model_->key_qvel[i * this->model_->nv], this->model_->nv);
+                    this->data_->time = this->model_->key_time[i];
+
+                    mju_copy(this->data_->ctrl, &this->model_->key_ctrl[i * this->model_->nu], this->model_->nu);
+                }
+            }
+
+            while (!this->stop_thread_) {
+                auto start_time = this->now();
+                {
+                    // Get the data from the shared vector safely
+                    std::lock_guard<std::mutex> lock(shared_data_mut_);
+                    for (int i = 0; i < nu_; i++) {
+                        data_->ctrl[i] = shared_data_.at(i);
+                    }
+                }
+
+                {
+                    std::lock_guard<std::mutex> lock(sensor_data_mut_);
+                    mj_step(model_, data_);
+                }
+
+                // Thread sleeping won't be precise enough for this, so instead we will spin lock the thread. This is
+                // inefficient, but accurate.
+                while ((this->now() - start_time).nanoseconds() < time_step_ * 1e9) {
+                }
+            }
+
+            RCLCPP_WARN_STREAM(this->get_logger(), "Cleaning up simulation data and model...");
+            // free MuJoCo model and data
+            mj_deleteData(data_);
+            mj_deleteModel(model_);
+
+            // rendering_thread.join();
```

Comment:
nit: delete this before merge, though you might want to join the thread in cleanup and/or shutdown explicitly, which would require moving the thread initialization out